### ISSUE TITLE
ConvertFrom-JSON high memory consumption

### ISSUE THREAD

**Issue by strawgate** [PowerShell/PowerShell#7698]
Steps to reproduce
------------------

```powershell
$JSON = @"
{
  "configGlossary:installationAt": "Philadelphia, PA",
  "configGlossary:adminEmail": "ksm@pobox.com",
  "configGlossary:poweredBy": "Cofax",
  "configGlossary:poweredByIcon": "\/images\/cofax.gif",
  "configGlossary:staticPath": "\/content\/static",
  "templateProcessorClass": "org.cofax.WysiwygTemplate",
  "templateLoaderClass": "org.cofax.FilesTemplateLoader",
  "templatePath": "templates",
  "templateOverridePath": "",
  "defaultListTemplate": "listTemplate.htm",
  "defaultFileTemplate": "articleTemplate.htm",
  "useJSP": false,
  "jspListTemplate": "listTemplate.jsp",
  "jspFileTemplate": "articleTemplate.jsp",
  "cachePackageTagsTrack": 200,
  "cachePackageTagsStore": 200,
  "cachePackageTagsRefresh": 60,
  "cacheTemplatesTrack": 100,
  "cacheTemplatesStore": 50,
  "cacheTemplatesRefresh": 15,
  "cachePagesTrack": 200,
  "cachePagesStore": 100,
  "cachePagesRefresh": 10,
  "cachePagesDirtyRead": 10,
  "searchEngineListTemplate": "forSearchEnginesList.htm",
  "searchEngineFileTemplate": "forSearchEngines.htm",
  "searchEngineRobotsDb": "WEB-INF\/robots.db",
  "useDataStore": true,
  "dataStoreClass": "org.cofax.SqlDataStore",
  "redirectionClass": "org.cofax.SqlRedirection",
  "dataStoreName": "cofax",
  "dataStoreDriver": "com.microsoft.jdbc.sqlserver.SQLServerDriver",
  "dataStoreUrl": "jdbc:microsoft:sqlserver:\/\/LOCALHOST:1433;DatabaseName=goon",
  "dataStoreUser": "sa",
  "dataStorePassword": "dataStoreTestQuery",
  "dataStoreTestQuery": "SET NOCOUNT ON;select test='test';",
  "dataStoreLogFile": "\/usr\/local\/tomcat\/logs\/datastore.log",
  "dataStoreInitConns": 10,
  "dataStoreMaxConns": 100,
  "dataStoreConnUsageLimit": 100,
  "dataStoreLogLevel": "debug",
  "maxUrlLength": 500
}
"@

$Count = 39999
$jsonarr = "[" + (($json + ",") * $Count) + $json + "]"
[system.gc]::Collect()
write-host ""
write-host ""
write-host "JSON -- Objects: $($Count + 1)"
write-host "JSON -- Length:  $($jsonarr.length / 1024 / 1024 ) Megabytes"

$BeforeMemory = $((get-process -pid $PID).WorkingSet)
[void] (convertfrom-json $jsonarr)
$AfterMemory = $((get-process -pid $PID).WorkingSet)

$MemoryDiff = $AfterMemory - $BeforeMemory

write-host "Memory Usage -- Before:     $($BeforeMemory  / 1024 / 1024) Megabytes"
write-host "Memory Usage -- After:      $($AfterMemory  / 1024 / 1024) Megabytes"
write-host "Memory Usage -- Difference: $($MemoryDiff  / 1024 / 1024) Megabytes"
write-host "Memory Usage -- Multiple:   $($MemoryDiff / $jsonarr.length)"

```
output:
```
JSON -- Objects: 40000
JSON -- Length:  67.4819955825806 Megabytes
Memory Usage -- Before:     210.8359375 Megabytes
Memory Usage -- After:      1240.90625 Megabytes
Memory Usage -- Difference: 1030.0703125 Megabytes
Memory Usage -- Multiple:   15.2643724241892
```
Actual behavior
-----------------
Powershell Core ends up with 1 GB of memory used and peaks to 1.6 GB of memory utilization during convertfrom-json

Expected behavior
-----------------

Powershell Core shouldn't need 1.6 gigabytes of memory to deserialize a 65MB JSON file.

Environment data
----------------

```powershell
Name                           Value
----                           -----
PSVersion                      6.1.0-preview.3
PSEdition                      Core
GitCommitId                    v6.1.0-preview.3
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by powercode** [comment]
We could consider interning property names.

**Comment by powercode** [comment]
Or, probably better, some name dictionary, were we resue the property names. That should help a lot with json memory consumption.

**Comment by markekraus** [comment]
This is actually a duplicate of #6199 as both commands use the same underlying code path.

**Comment by iSazonov** [comment]
The issue is still present.

/cc @david-tap

**Comment by iSazonov** [comment]
After the code review I think we need to  migrate to System.Text.Json to get performance benefits.

**Comment by jszabo98** [comment]
Note that convertfrom-json has no streaming mode like jq.  So it will load the whole file into memory, no matter how big it is.

**Comment by hanneshayashi** [comment]
> Note that convertfrom-json has no streaming mode like jq. So it will load the whole file into memory, no matter how big it is.

Would be great if it had a streaming mode, like jq, though...

**Comment by iSazonov** [comment]
We would need support from System.Text.Json .Net API. Could you open new issue in .Net Runtime repository for this streaming scenario?

**Comment by jasoth** [comment]
It sounds like the streaming functionality might exist now in the System.Text.Json APIs.
https://devblogs.microsoft.com/dotnet/whats-next-for-system-text-json/

It would be great to have pipeline friendly JSON commands for large datasets including an Export-Json/Import-Json for reading and writing directly to and from disk.

**Comment by iSazonov** [comment]
@jasoth You can open new issue with your proposal. Please describe scenarios you want to address. Voting from community is important.

**Comment by strawgate** [comment]
I ended up writing a module called ConvertFrom-JsonFast that implements system.text.json here: https://github.com/strawgate/ConvertFrom-JsonFast or on the gallery here https://www.powershellgallery.com/packages/ConvertFrom-JsonFast/1.0.0

1000 MB JSON File
```
ConvertFrom-JsonFast:           5950 MB used over 16400 Milliseconds
ConvertFrom-Json:              13575 MB used over 111309 Milliseconds
```

100 MB JSON File
```
ConvertFrom-JsonFast:           544 MB used over 1400 Milliseconds
ConvertFrom-Json:              1330 MB used over 9429 Milliseconds
```

**Comment by SteveL-MSFT** [comment]
@strawgate that's some pretty impressive results!  Do you know if your version would pass all the existing tests that we have for `ConvertFrom-Json`

**Comment by iSazonov** [comment]
> Do you know if your version would pass all the existing tests that we have for ConvertFrom-Json

This is absolutely impossible since system.text.json implementation has a long list of incompatibilities (by design).

My suggestion was and is to implement new cmdlets based on system.text.json and leave the old ones so that users can switch to them if necessary.

**Comment by strawgate** [comment]
I doubt it'll pass all the tests but knowing which tests it fails would be useful! I'll take a look at how that might work!

The goal was to be a drop-in replacement for simple use cases but more complex ones will likely have different behavior.

**Comment by iSazonov** [comment]
See #11198 for information.

**Comment by strawgate** [comment]
I know this is the wrong place to talk about the module but i'll just drop one more note:

I ran the Json tests against ConvertFrom-JsonFast and found a couple of issues that I resolved. At this time i have 216/224 tests passing with the only tests failing being tests that check for the exact exception type being thrown.

While the ConvertFrom-Json test coverage isn't great, it's nice to see that it could be fairly compatible!
